### DISCUSSION TITLE
CMake Build Error

### DISCUSSION THREAD

**Discussion by BssnakeMing** [obsproject/obs-studio#10644] [Development Questions]
`F:\CppProject\obs-studio>cmake --preset windows-x64
Preset CMake variables:

  ENABLE_BROWSER:BOOL="TRUE"
  ENABLE_CCACHE:BOOL="FALSE"
  GPU_PRIORITY_VAL:STRING=""
  OBS_CMAKE_VERSION:STRING="3.0.0"
  RESTREAM_CLIENTID:STRING=""
  RESTREAM_HASH:STRING=""
  TWITCH_CLIENTID:STRING=""
  TWITCH_HASH:STRING=""
  VIRTUALCAM_GUID:STRING="A3FCE0F5-3493-419F-958A-ABA1250EC20B"
  YOUTUBE_CLIENTID:STRING=""
  YOUTUBE_CLIENTID_HASH:STRING=""
  YOUTUBE_SECRET:STRING=""
  YOUTUBE_SECRET_HASH:STRING=""

-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.32.31340.0
-- The CXX compiler identification is MSVC 19.32.31340.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: D:/Microsoft/Visual Studio/2022/VC/Tools/MSVC/14.32.31326/bin/HostX64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: D:/Microsoft/Visual Studio/2022/VC/Tools/MSVC/14.32.31326/bin/HostX64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting up Pre-Built obs-deps (x64)
-- Setting up Pre-Built obs-deps (x64) - skipped
-- Setting up Pre-Built Qt6 (x64)
-- Setting up Pre-Built Qt6 (x64) - skipped
-- Setting up Chromium Embedded Framework (x64)
-- Setting up Chromium Embedded Framework (x64) - skipped
-- Selecting Windows SDK version 10.0.22621.0 to target Windows 10.0.19045.
-- The C compiler identification is MSVC 19.32.31340.0
-- The CXX compiler identification is MSVC 19.32.31340.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: D:/Microsoft/Visual Studio/2022/VC/Tools/MSVC/14.32.31326/bin/HostX64/x86/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: D:/Microsoft/Visual Studio/2022/VC/Tools/MSVC/14.32.31326/bin/HostX64/x86/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Setting up Pre-Built obs-deps (x86)
-- Setting up Pre-Built obs-deps (x86) - skipped
CMake Warning (dev) at cmake/finders/FindDetours.cmake:81 (message):
  Failed to find detours version.
Call Stack (most recent call first):
  plugins/win-capture/graphics-hook/CMakeLists.txt:3 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

-- Configuring done (6.7s)
-- Generating done (0.0s)
-- Build files have been written to: F:/CppProject/obs-studio/build_x86
CMake Warning (dev) at cmake/finders/FindFFmpeg.cmake:225 (message):
  Failed to find avformat version.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:148 (ffmpeg_check_version)
  cmake/finders/FindFFmpeg.cmake:264 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/finders/FindFFmpeg.cmake:225 (message):
  Failed to find avutil version.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:148 (ffmpeg_check_version)
  cmake/finders/FindFFmpeg.cmake:264 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/finders/FindFFmpeg.cmake:225 (message):
  Failed to find swscale version.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:148 (ffmpeg_check_version)
  cmake/finders/FindFFmpeg.cmake:264 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/finders/FindFFmpeg.cmake:225 (message):
  Failed to find swresample version.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:148 (ffmpeg_check_version)
  cmake/finders/FindFFmpeg.cmake:264 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/finders/FindFFmpeg.cmake:225 (message):
  Failed to find avcodec version.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:148 (ffmpeg_check_version)
  cmake/finders/FindFFmpeg.cmake:264 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Warning (dev) at cmake/finders/FindFFmpeg.cmake:225 (message):
  Failed to find avutil version.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:148 (ffmpeg_check_version)
  cmake/finders/FindFFmpeg.cmake:275 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at cmake/finders/FindFFmpeg.cmake:177 (cmake_path):
  cmake_path undefined variable for input path.
Call Stack (most recent call first):
  cmake/finders/FindFFmpeg.cmake:156 (ffmpeg_find_dll)
  cmake/finders/FindFFmpeg.cmake:275 (ffmpeg_find_component)
  libobs/CMakeLists.txt:16 (find_package)

CMake Error at C:/Program Files/CMake/share/cmake-3.29/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find FFmpeg (missing: FFmpeg_LIBRARIES FFmpeg_INCLUDE_DIRS
  avformat avutil swscale swresample) (Required is at least version "6")

      Reason given by package: Ensure that obs-deps is provided as part of CMAKE_PREFIX_PATH.

Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.29/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  cmake/finders/FindFFmpeg.cmake:294 (find_package_handle_standard_args)
  libobs/CMakeLists.txt:16 (find_package)

-- Configuring incomplete, errors occurred!`

What do I need to do?

**Comment by OnlyDelusion** [comment]
I've solved a similar problem, and you may have seen some old compilation tutorials like me. There is indeed something wrong with this compilation document. Especially when operating manually.
In the new version, you only need to set 3 paths in the Cmake GUI, 
1. QT's msvc path 
2. Dependent package path. 
3. Generated file saving path. (new version only set CMAKE_PREFIX_PATH)
The problem here is the first dependency package path，we need to download four file( cef_binary_5060_windows_x64.zip obs-deps-2024-05-08-x64.zip obs-deps-2024-05-08-x86.zip obs-deps-qt6-2024-05-08-x64.zip) and They should be placed under this path (xxx/obs-studio/.deps/)  and then Add the variable CMAKE_PREFIX_PATH=xxx/obs-studio/.deps/ in the cmake-gui interface。 

Then I compiled successfully
1.git clone --recursive https://github.com/obsproject/obs-studio.git
2.cd obs-studio
3.cmake --preset windows-x64
4.run cmake-gui
5. Add the variable CMAKE_PREFIX_PATH=xxx/obs-studio/.deps/ in the cmake-gui interface。
6.  Add the variable QTDIR=xxx/QT/msvc2022_64/ in the cmake-gui interface。
7. Add the variable CMAKE_INSTALL_PERFIX= the path you want to save in the cmake-gui interface。
8.config generate and open ，and then to build。

**Reply by RytoEX** [reply]
You don't have to manually download CEF and you don't have to do steps 4-7. If you use one of our predefined CMake presets, CMake will automatically download our prebuilt dependencies for you and set the CMAKE_PREFIX_PATH accordingly. You only need steps 1-3 and 8 if you are not using custom dependencies.
